theano/tensor/nnet/nnet.py

```diff
@@ -889,6 +890,39 @@ def local_logsoftmax(node):
         return [ret]
 
 
+@opt.register_specialize('fast_compile')
+@gof.local_optimizer([tensor.Elemwise])
+def local_indexing_logsoftmax(node):
+    list_classes = ((Subtensor, AdvancedSubtensor1, AdvancedSubtensor))
+    if (isinstance(node.op, tensor.Elemwise) and
+            isinstance(node.op.scalar_op, scalar.basic.Log) and
+            node.inputs[0].owner is not None):
+
+            if isinstance(node.inputs[0].owner.op, Softmax):
+                softmax_input = node.inputs[0].owner.inputs[0]
+                axis = node.inputs[0].owner.op.axis
+                ret = LogSoftmax(axis)(softmax_input)
+                ret.tag.values_eq_approx = values_eq_approx_remove_inf
+                copy_stack_trace([node.inputs[0], node.outputs[0]], ret)
+                return [ret]
+
+            elif isinstance(node.inputs[0].owner.op, list_classes):
+                subtensor_op = node.inputs[0].owner.op
+                subtensor_input = node.inputs[0].owner.inputs[0]
+                subtensor_inputs = node.inputs[0].owner.inputs[1]
+                if isinstance(subtensor_input.owner.op, Softmax):
+                    softmax_op = subtensor_input.owner.op
+                    softmax_input = subtensor_input.owner.inputs[0]
+                    axis = softmax_op.axis
+                    ret = subtensor_op(LogSoftmax(axis)(softmax_input), subtensor_inputs)
+                    ret.tag.values_eq_approx = values_eq_approx_remove_inf
+                    copy_stack_trace([node.inputs[0], node.outputs[0]], ret)
+                    return [ret]
+
+            else:
```

Comment:
At the end of such a function it will return any way so no ```else return``` is needed. Its existence does not hurt any logics but let's remove it for the sake of consistency.